File: Olympiad/COCI/official/2009/regional_solutions/reljef.cpp
#include <algorithm>
#include <cstdio>
#include <vector>

using namespace std;

typedef pair<int,int> par;

#define MAXR 100
#define MAXS 100

int R, S;
char a[MAXR][MAXS+1];
char bio[MAXR][MAXS+1];

vector<par> nakupina;

void dfs( int r, int s ) {
   if( r < 0 || r >= R ) return;
   if( s < 0 || s >= S ) return;
   if( a[r][s] == '.' ) return;
   if( bio[r][s] ) return;
   bio[r][s] = 1;
   nakupina.push_back( par(r,s) );
   dfs( r, s-1 );
   dfs( r, s+1 );
   dfs( r-1, s );
   dfs( r+1, s );
}

void srusi() {
   memset( bio, 0, sizeof bio );

   for( int r = 0; r < R; ++r ) {
      for( int s = 0; s < S; ++s ) {
         if( a[r][s] == '.' || bio[r][s] ) continue;
         
         nakupina.clear();
         dfs( r, s );

         vector<int> najniza( S, -1 );
         for( vector<par>::iterator it = nakupina.begin(); it != nakupina.end(); ++it ) {
            najniza[it->second] = max( najniza[it->second], it->first );
            a[it->first][it->second] = '.';
         }

         int padni = R;
         for( int i, j = 0; j < S; ++j ) {
            if( najniza[j] == -1 ) continue;
            for( i = najniza[j]+1; i < R && a[i][j] == '.'; ++i );
            padni = min( padni, i-najniza[j]-1 );
         }
         
         for( vector<par>::iterator it = nakupina.begin(); it != nakupina.end(); ++it ) {
            it->first += padni;
            a[it->first][it->second] = 'x';
            bio[it->first][it->second] = 1;
         }
      }
   }
}

int main( void ) {
   scanf( "%d%d", &R, &S );
   for( int r = 0; r < R; ++r ) scanf( "%s", a[r] );

   int N;
   scanf( "%d", &N );
   for( int i = 0; i < N; ++i ) {
      int h;
      scanf( "%d", &h );

      int r = R-h;
      if( i % 2 == 0 ) {
         for( int s = 0; s < S; ++s )
            if( a[r][s] == 'x' ) { a[r][s] = '.'; break; }
      } else {
         for( int s = S-1; s >= 0; --s )
            if( a[r][s] == 'x' ) { a[r][s] = '.'; break; }
      }
      
      srusi();
   }

   for( int r = 0; r < R; ++r ) printf( "%s\n", a[r] );
   return 0;
}

File: Olympiad/COCI/official/2009/regional_solutions/reljef.pas
program reljef;

const MAXR = 100;
const MAXS = 100;

var 
	R, S, N, i, j, k: longint;
	a : array[1..MAXR,1..MAXS+1] of char;
   bio : array[1..MAXR,1..MAXS] of boolean;

   n_nakupina : longint;
   nakupina_r : array[1..MAXR*MAXS] of longint;
   nakupina_s : array[1..MAXR*MAXS] of longint;

function min( a, b : longint ) : longint;
begin
   if a < b then min := a else min := b;
end;
function max( a, b : longint ) : longint;
begin
   if a > b then max := a else max := b;
end;
   
procedure dfs( i, j : longint );
begin
   if (i < 1) or (i > R) then exit;
   if (j < 1) or (j > S) then exit;
   if a[i, j] = '.' then exit;
   if bio[i, j] then exit;
   bio[i, j] := true;
   
   n_nakupina := n_nakupina + 1;
   nakupina_r[n_nakupina] := i;
   nakupina_s[n_nakupina] := j;
   
   dfs( i-1, j );
   dfs( i+1, j );
   dfs( i, j-1 );
   dfs( i, j+1 );
end;

procedure srusi;
var
   i, j, k, rr, ss, padni : longint;
   najniza : array[1..MAXS] of longint;
begin
   for i := 1 to R do
      for j := 1 to S do
         bio[i,j] := false;

   for i := 1 to R do begin
      for j := 1 to S do begin
         if (a[i,j] = '.') or bio[i,j] then continue;
         
         n_nakupina := 1;
         dfs( i, j );

         for ss := 1 to S do najniza[ss] := 0;
         
         for k := 1 to n_nakupina do begin
            najniza[ nakupina_s[k] ] := max( najniza[ nakupina_s[k] ], nakupina_r[k] );
            a[nakupina_r[k], nakupina_s[k]] := '.';
         end;

         padni := R;         
         for ss := 1 to S do begin            
            if najniza[ss] = 0 then continue;
            rr := najniza[ss] + 1;
            while (rr <= R) and (a[rr,ss] = '.') do rr := rr + 1;           
            padni := min( padni, rr-najniza[ss]-1 );
         end;
         
         for k := 1 to n_nakupina do begin
            nakupina_r[k] := nakupina_r[k] + padni;
            a[nakupina_r[k], nakupina_s[k]] := 'x';
            bio[nakupina_r[k], nakupina_s[k]] := true;
         end;
      end;
   end;
end;

begin
   readln( R, S );
   for i := 1 to R do readln( a[i] );   

   readln( N );
   for k := 1 to N do begin
      read( i );
      i := R-i+1;      

      if k mod 2 = 1 then begin
         for j := 1 to S do         
            if a[i, j] = 'x' then begin
               a[i, j] := '.'; 
               break;
            end;
      end else begin
         for j := S downto 1 do 
            if a[i, j] = 'x' then begin
               a[i, j] := '.'; 
               break;
            end;
      end;
      
      srusi;
   end;

   for i := 1 to R do begin
      for j := 1 to S do
         write( a[i,j] );
      writeln;
   end;
end.